### PR TITLE
bcm270x: Convert to ARCH_MULTIPLATFORM

### PR DESCRIPTION
**Pull request by notro**
This PR converts ARCH_BCM2708 and ARCH_BCM2709 to ARCH_MULTIPLATFORM.

Both ARCH_BCM2708 and ARCH_BCM2709 are now based on ARCH_BCM2835, removing the need for the mach-bcm2708 and mach-bcm2709 directories. In addition to that it removes the need to patch many Kconfig files just to add `depends ARCH_BCM2708 || ARCH_BCM2708 || ...`.

Using ARCH_BCM2835 means that the 2835 dtb's are also built, but they won't work with this config.
And scripts/mkknlimg will add the tag: `283x: y`

Tested on Pi1 and Pi2.

@popcornmix please check with the GPU debugger to make sure that this doesn't give `L2 bad address exception` like it did when I tried this last year: https://github.com/raspberrypi/linux/pull/1178#issuecomment-153525832

Here's the .config diff introduced by this PR:
- 2708: https://gist.github.com/notro/e0b8f6997f29afa438724be629abc848
- 2709: https://gist.github.com/notro/18a784e2707c7082571404058bb1d7a3

Of all the PR's I have submitted, this is the one I have the least confidence in. I have had so much problems and lockups working with this ARCH_MULTIPLATFORM convertion, both now and last year, that I started to think that I just couldn't do it.
That being said, I haven't had any issues with this particular PR, just with all my efforts to try and build a stable kernel that can boot on both Pi1 and Pi2.

### CODE REVIEW

**Comment by notro** [comment]
**Unified kernel image**

It is now possible to build a kernel image that can/should boot on Pi1 and Pi2.
I based the kernel on `bcm2709_defconfig` with these changes:

```
CONFIG_ARCH_MULTI_V6=y
CONFIG_ARCH_BCM2708=y
# CONFIG_MACH_BCM2708 is not set
# CONFIG_NEON is not set
```

It seems to work on Pi1 except that `vchiq_test -p` doesn't work, but `vchiq_test -f` does.

On Pi2 it isn't good. Tagging the kernel from a network driver always fails like this:

```
$ mount -t nfs 192.168.10.92:/home/pi /home/pi/work -o rw
$ ~/work/notro-raspberrypi-linux/notro-raspberrypi-linux/scripts/mkknlimg --dtok ~/work/notro-raspberrypi-linux/notro-raspberrypi-linux/arch/arm/boot/zImage ~/kernel.img
Illegal instruction
```

And every other boot drops me into emergency console (eventually):

```
[  *** ] (1 of 1) A start job is running for dev-mmcblk0p1.de... 1s / 1min 30s)
<...>
[ TIME ] Timed out waiting for device dev-mmcblk0p1.device.
[DEPEND] Dependency failed for /boot.
[DEPEND] Dependency failed for Local File Systems.
[DEPEND] Dependency failed for File System Check on /dev/mmcblk0p1.

Welcome to emergesulogin: root account is locked, starting shell
root@raspberrypi:~#
```

I had this problem last year also when I tried this, and I haven't figured out the cause.
I have tried switching to dwc2 and disabling UACCESS_WITH_MEMCPY and BCM2708_VCHIQ.

Oh well, I wasn't able to make that unified image, but hopefully this PR survives review and testing, with the result that downstream for all pratical purposes will be using ARCH_BCM2835!

**Comment by popcornmix** [comment]
Looks good.
@P33M dwc_otg commit looks okay to me - any objections?

**Comment by pelwell** [comment]
One interesting aspect of having 270x and 283x support in the same image is that the mkknlimg trailer (if you add one) sets both flags. The firmware currently sees the 283x flag and chooses to load bcm283x dtb files. If they aren't there, as is the case with my (remote) Pi, it won't boot (not falling back to the bcm270x variants). Omitting the trailer should get it working again.

**Comment by notro** [comment]
I see now that there are two flags: 283x and 283X. But in my case it loads the right dtb:

```
$ ~/work/notro-raspberrypi-linux/notro-raspberrypi-linux/scripts/knlinfo /boot/kernel.img
Kernel trailer found at 4341168/0x423db0:
  KVer: "Linux version 4.8.4+ (pi@raspi2) (gcc version 4.8.3 20140106 (prerelease) (crosstool-NG linaro-1.13.1-4.8-2014.01 - Linaro GCC 2013.11) ) #1 Tue Oct 25 19:24:32 CEST 2016"
  DTOK: y
  DDTK: y
  270X: y
  283X: y
  283x: n

$ sudo vcdbg log msg
001413.403: Loading 'kernel.img' to 0x8000 size 0x423eb4
001417.054: Kernel trailer DTOK property says yes
001417.068: Kernel trailer DDTK property says yes
001419.353: Loading 'bcm2708-rpi-b-plus.dtb' to 0x42beb4 size 0x31a7

```

**Comment by pelwell** [comment]
That was a false alarm - "283x" is the old single-platform tag, and if set will result in the bcm283x DTBs being loaded . "283X" says the image includes upstream support, but not to use it automatically because it also includes "270X" support ("283x" means "283X && !270X").

On a Pi 2 I get complaints about timer interrupts:

```
[    0.000000] Hierarchical RCU implementation.
[    0.000000]  Build-time adjustment of leaf fanout to 32.
[    0.000000] NR_IRQS:16 nr_irqs:16 16
[    0.000000] arm_arch_timer: WARNING: Invalid trigger for IRQ16, assuming level low
[    0.000000] arm_arch_timer: WARNING: Please fix your firmware
[    0.000000] arm_arch_timer: WARNING: Invalid trigger for IRQ17, assuming level low
[    0.000000] arm_arch_timer: WARNING: Please fix your firmware
[    0.000000] arm_arch_timer: Architected cp15 timer(s) running at 19.20MHz (phys).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0x46d987e47, max_idle_ns: 440795202767 ns
[    0.000010] sched_clock: 56 bits at 19MHz, resolution 52ns, wraps every 4398046511078ns
[    0.000030] Switching to timer-based delay loop, resolution 52ns
[    0.000377] Console: colour dummy device 80x30
[    0.001872] console [tty1] enabled
[    0.001927] Calibrating delay loop (skipped), value calculated using timer frequency.. 38.40 BogoMIPS (lpj=192000)
[    0.002004] pid_max: default: 32768 minimum: 301
[    0.002419] Mount-cache hash table entries: 2048 (order: 1, 8192 bytes)
[    0.002473] Mountpoint-cache hash table entries: 2048 (order: 1, 8192 bytes)
[    0.003762] Disabling cpuset control group subsystem
[    0.004028] CPU: Testing write buffer coherency: ok
[    0.004124] ftrace: allocating 21951 entries in 65 pages
[    0.058851] CPU0: update cpu_capacity 1024
[    0.058932] CPU0: thread -1, cpu 0, socket 15, mpidr 80000f00
[    0.059031] Setting up static identity map for 0x100000 - 0x100034
[    0.062111] arm_arch_timer: WARNING: Invalid trigger for IRQ16, assuming level low
[    0.062117] arm_arch_timer: WARNING: Please fix your firmware
[    0.062129] arm_arch_timer: WARNING: Invalid trigger for IRQ17, assuming level low
[    0.062133] arm_arch_timer: WARNING: Please fix your firmware
[    0.062151] CPU1: update cpu_capacity 1024
[    0.062159] CPU1: thread -1, cpu 1, socket 15, mpidr 80000f01
[    0.063174] arm_arch_timer: WARNING: Invalid trigger for IRQ16, assuming level low
[    0.063179] arm_arch_timer: WARNING: Please fix your firmware
[    0.063186] arm_arch_timer: WARNING: Invalid trigger for IRQ17, assuming level low
[    0.063189] arm_arch_timer: WARNING: Please fix your firmware
[    0.063203] CPU2: update cpu_capacity 1024
[    0.063210] CPU2: thread -1, cpu 2, socket 15, mpidr 80000f02
[    0.064127] arm_arch_timer: WARNING: Invalid trigger for IRQ16, assuming level low
[    0.064132] arm_arch_timer: WARNING: Please fix your firmware
[    0.064139] arm_arch_timer: WARNING: Invalid trigger for IRQ17, assuming level low
[    0.064142] arm_arch_timer: WARNING: Please fix your firmware
[    0.064156] CPU3: update cpu_capacity 1024
[    0.064164] CPU3: thread -1, cpu 3, socket 15, mpidr 80000f03
[    0.064273] Brought up 4 CPUs
[    0.064880] SMP: Total of 4 processors activated (153.60 BogoMIPS).
```

My Pi was failing to boot because it's a Pi3, and both USB and the UART seem to be hosed with this build. More digging required.

**Comment by notro** [comment]
#1669 has info about the warnings.

**Comment by pelwell** [comment]
vchiq isn't connecting. Still digging.

**Comment by popcornmix** [comment]
I have the same problem. Build with bcm2709_defconfig works on Pi2.
On Pi3 usb doesn't work. Hub is detected but devices aren't:
https://www.dropbox.com/s/ud8g277qnu41ktq/IMG_1718.JPG?dl=0

**Comment by P33M** [comment]
I see no issues with the dwc_otg changes.

**Comment by pelwell** [comment]
The problem was caused by the lack of a dma-ranges property in the bcm2710 dtb. I've added a patch that includes it, and the `reg` property for the `intc` node which was also missing.

File: arch/arm/mach-bcm/Kconfig [hunk 1]
```diff
@@ -160,6 +160,20 @@ config ARCH_BCM2835
 	  This enables support for the Broadcom BCM2835 and BCM2836 SoCs.
 	  This SoC is used in the Raspberry Pi and Roku 2 devices.
 
+config MACH_BCM2708
+	bool "Enable optimized __copy_to_user and __copy_from_user"
```
Comment (popcornmix) [inline arch/arm/mach-bcm/Kconfig:164]: This text doesn't seem to match the config name. A c/p error?

Comment (notro) [reply arch/arm/mach-bcm/Kconfig:164]: I was tired and didn't know what to call it, so I just kept the name.
The only place CONFIG_MACH_BCM2708 is used is in the optimization code: https://github.com/raspberrypi/linux/commit/5fb515e14129b40fa8f3d8bf797d0a80b82fe512

**Comment by notro** [comment]
Thanks for fixing that up @pelwell, I've been down so many rabbit holes with this, that I forgot about 2710.

**Comment by popcornmix** [comment]
FYI I've updated 4.9 tree with these changes, rebasing them into original commits.
Looks quite a lot cleaner. :-)

**Comment by popcornmix** [comment]
@pelwell okay to merge?

**Comment by notro** [comment]
Yes the Main bcm2708/bcm2709 linux port commit is significantly smaller now!

I have some i2c-bcm2835 [patches](http://marc.info/?l=linux-i2c&m=147552520707011&w=4) that awaits i2c maintainer review. With those in place, I hope we can switch to use i2c-bcm2835 by default. AFAIK that's the last of the mainline drivers which has a downstream 2708 version.

**Comment by popcornmix** [comment]
@notro great - we always look forward to your PR's - keep them coming!

**Comment by ghost** [comment]
As a Raspberry Pi enthusiast I love threads like this, despite understanding only the most general aspects. Thanks to all involved, and please keep up the great work.

**Comment by pelwell** [comment]
@notro Don't forget the bcm2708-usb - we'll be holding on to that for a while, I think. And thank you for another significant contribution.

Yes - merge away.

**Comment by notro** [comment]
Yes you're right of course, I forgot that one. I had a look at the dwc_otg patch and it touches core files for the fiq support, so it will probably stay a long time. IIRC there was a comment in an arm64 thread about dwc_otg outperforming dwc2 even without fiq.

**Comment by P33M** [comment]
The main hurdles to dwc_otg and associated tentacles going away are:
- FIQ support is required on AARCH64 for the dwc host controller to have any hope of operating reliably
- We need a better way of making sure FIQs are always unmasked on the CPU that's servicing FIQs
- Transplanting the FIQ state machine and associated communication mechanisms across to a driver with substantially different data structures and locking is going to introduce an inordinate number of bugs/edge cases
- Incompatibility with other SoC implementations - there's a lot of hardcoded yuk used to get the FIQ to wakeup an IRQ - which most certainly won't work on other platforms.

Gnarly but not insurmountable (for Herculean quantities of time | effort).

Also @ncguk : Stick around, you might learn a thing or two.

**Comment by clivem** [comment]
Related to these MULTI_PLATFORM changes?????

```
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_set_clock:84: Failed to set clock: 600000 (-12)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_driver_target_index:183: Error occurred setting a new frequency (600000)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: cpufreq: __target_index: Failed to change cpu frequency: -22
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_set_clock:84: Failed to set clock: 600000 (-12)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_driver_target_index:183: Error occurred setting a new frequency (600000)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: cpufreq: __target_index: Failed to change cpu frequency: -22
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_set_clock:84: Failed to set clock: 600000 (-12)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_driver_target_index:183: Error occurred setting a new frequency (600000)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: cpufreq: __target_index: Failed to change cpu frequency: -22
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_set_clock:84: Failed to set clock: 600000 (-12)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_driver_target_index:183: Error occurred setting a new frequency (600000)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: cpufreq: __target_index: Failed to change cpu frequency: -22
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_set_clock:84: Failed to set clock: 600000 (-12)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: bcm2835-cpufreq:bcm2835_cpufreq_driver_target_index:183: Error occurred setting a new frequency (600000)
Oct 28 23:15:04 MamboLSPlus2B1 kernel: cpufreq: __target_index: Failed to change cpu frequency: -22
```

**Comment by ghost** [comment]
Hi, someone sent me a link to this PR.

I've seen very similar clock switches issues on the RPI 3 running in aarch64 mode.  Does anybody have any ideas what the issue is?

I initially thought it had something to do with the power supply, but after beefing that up I still see it occasionally.  I also noticed that recent kernel changes to 8.1 and 8.2 features seem to have made differences as well.   I was able to track these issues down to weird hangs in the firmware.

**Comment by notro** [comment]
-12 is -ENOMEM.

Call chain:
bcm2835_cpufreq_set_clock() -> bcm2835_cpufreq_clock_property() -> rpi_firmware_property() -> rpi_firmware_property_list() 

```
int rpi_firmware_property_list(struct rpi_firmware *fw,
                   void *data, size_t tag_size)
{

    buf = dma_alloc_coherent(fw->cl.dev, PAGE_ALIGN(size), &bus_addr,
                 GFP_ATOMIC);
    if (!buf)
        return -ENOMEM;
```

Try this:

/boot/cmdline.txt

```
coherent_pool=4M 
```

```
$ dmesg | grep prealloc
[    0.298262] DMA: preallocated 4096 KiB pool for atomic coherent allocations
```

This value was changed in #1678 from 4096kB to the default 256kB, which mainline also uses.

**Comment by notro** [comment]
@clivem Please try this patch so we can find out who's eating your atomic dma pool:

``` diff
diff --git a/arch/arm/mm/dma-mapping.c b/arch/arm/mm/dma-mapping.c
index c6834c0..4739b68 100644
--- a/arch/arm/mm/dma-mapping.c
+++ b/arch/arm/mm/dma-mapping.c
@@ -714,6 +714,11 @@ static struct arm_dma_allocator cma_allocator = {
 static void *pool_allocator_alloc(struct arm_dma_alloc_args *args,
                                  struct page **ret_page)
 {
+   static int i;
+
+   if (i++ < 100)
+       pr_info("%s: size=%zu, caller=%ps (avail=%zu)\n", __func__, args->size, args->caller, gen_pool_avail(atomic_pool));
+
        return __alloc_from_pool(args->size, ret_page);
 }

```

The 100 cap is to avoid coming to a crawl if there's a _lot_ of allocations.

This is my Pi2 right after boot:

```
pi@raspberrypi:~$ dmesg | grep caller
[    0.478049] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    0.760758] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    0.773412] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    0.773728] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    0.779104] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    0.837454] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    1.919442] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    1.962206] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    2.690793] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    2.715092] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    2.729154] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    2.751496] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    2.766190] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    2.985962] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    3.025659] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    3.068247] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    3.161925] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    3.283724] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    8.881481] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    8.882480] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
[    9.820790] pool_allocator_alloc: size=65536, caller=__DWC_DMA_ALLOC_ATOMIC (avail=262144)
[    9.848402] pool_allocator_alloc: size=65536, caller=__DWC_DMA_ALLOC_ATOMIC (avail=196608)
[   11.708326] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=131072)
[   17.699811] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=131072)
[   18.428451] pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=131072)
```

**Comment by clivem** [comment]
```
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:02 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:14 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:14 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:19 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:19 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=262144)
Feb 11 16:28:20 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=65536, caller=__DWC_DMA_ALLOC_ATOMIC (avail=262144)
Feb 11 16:28:20 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=196608)
Feb 11 16:28:20 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=196608)
Feb 11 16:28:20 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=65536, caller=__DWC_DMA_ALLOC_ATOMIC (avail=196608)
Feb 11 16:28:22 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=131072)
Feb 11 16:28:22 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=131072)
Feb 11 16:28:23 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=131072)
Feb 11 16:28:23 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=131072)
Feb 11 16:28:23 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=65536, caller=__DWC_DMA_ALLOC_ATOMIC (avail=131072)
Feb 11 16:28:23 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=65536, caller=__DWC_DMA_ALLOC_ATOMIC (avail=65536)
Feb 11 16:28:24 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=0)
Feb 11 16:28:26 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=0)
Feb 11 16:28:26 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=0)
Feb 11 16:28:26 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=0)
Feb 11 16:28:26 MamboLSPlus2B1 kernel: pool_allocator_alloc: size=4096, caller=rpi_firmware_property_list (avail=0)

```

**Comment by clivem** [comment]
@notro Sorry, I'm being a retard again. Of course, I already "fixed" (worked around) this with coherent_pool=4M, (to get back to the previous default), the last time I brought it up, with your first set of platform cleanup patches. https://github.com/raspberrypi/linux/pull/1678#commitcomment-19396961  That default does need increasing, me thinks. Maybe not back to 4M, but more than 256k. ;)

**Comment by notro** [comment]
Thanks @clivem, the usb driver is eating up your pool.

@P33M How much atomic dma memory does dwc_otg need?

**Comment by P33M** [comment]
https://github.com/raspberrypi/linux/blob/24e62728b3fc4f118c8ae17b374bce189bb188fc/drivers/usb/host/dwc_otg/dwc_otg_hcd.c#L1375

This is why - when an endpoint gets a URB submitted that doesn't have a 32-bit aligned start address, a swizzle buffer is allocated that's 64k in size. This buffer persists until the endpoint is shutdown or the device is removed.

Most drivers play "nice" - regular kmalloc calls will return sensible alignments - but certain devices do hacky partitioning of their own buffer pools which results in this behaviour.

**Comment by ghost** [comment]
@notro: Thanks for the explanation about the DMA pool size.  It makes sense and I understand better now.

Sorry, to chim in on this but I just recently thought of an issue with a unified binary.

In theory: PI2 can use the standard GNU arm-linux-gnueabihf- compiler.  In fact that's what I do when building a PI2 kernel myself.  But I understand that PI1 needs either the non hardware floating point compiler or the foundation's custom gcc compiler.  

So I'm wondering to myself if I even makes sense to have a unified kernel binary when PI2 can benefit from the standard gnu gcc release.

**Comment by ghost** [comment]
FYI, Before I get flamed too bad here.  I think cleaning up the configuration environment and makes the configurations as similar as possible is a really cool idea.

All I'm saying is that perhaps it still makes sense for the actual binary to be a different image because the PI2 or PI3 can take advantage of a different compiler or compiler switches.

**Comment by notro** [comment]
> So I'm wondering to myself if I even makes sense to have a unified kernel binary

No not really since we loose performance: the optimized __copy_to_user and __copy_from_user on Pi1 and CONFIG_NEON on Pi2.
My interest was just that it **should** be possible to build such a kernel image, and I don't like to give up and not knowing what the problem is.

**Comment by notro** [comment]
Ok, so the default 256kB is not enough. I don't know how big it should be, 4MB has shown itself to be enough, but I don't think this pool can be used in OOM situations, so it shouldn't be too big.
On Pi2 and Pi3 memory is plentiful, so it's Pi1 that's really affected by this. I'll leave this for others to decide.

arch/arm/mach-bcm/board_bcm2835.c

``` c
#include <linux/dma-mapping.h>

static void __init bcm2835_init_early(void)
{
    /* dwc_otg needs this for bounce buffers on non-aligned transfers */
    init_dma_coherent_pool_size(SZ_4M);
}

MACHINE_START(...)
    .init_early = bcm2835_init_early,
MACHINE_END
```

**Comment by clivem** [comment]
I chuckled earlier today, when I came across this from Dec 2012.... https://github.com/raspberrypi/linux/issues/170

**Comment by clivem** [comment]
BTW, I've been using "coherent_pool=1M" on the 2x Zero's, 4x 2B's and 3x 3B's, that I've now got running 4.8.y HEAD with ARCH_MULTIPLATFORM config. Perhaps that is a sensible compromise between 256k and 4M?

**Comment by popcornmix** [comment]
@clivem #170 was before https://github.com/raspberrypi/linux/commit/67fd2c5af2cdfa7a1659258e0f4f7ec8269a0dd6 so coherent memory requirements are less now then they once were.
However I agree it needs increasing, or dwc_otg needs to reduce its requirement. We'll make a decision about this tomorrow.

**Comment by clivem** [comment]
@popcornmix Yes.... Understand..... I just chuckled at the headline, "256 KiB atomic DMA coherent pool is too small!", 4 years ago, rather than what was causing it to be too small or where it fits in terms of timelines of what happened, when, and other commits.

**Comment by ghost** [comment]
OK, silly question here.  I'm new to this whole thing....

Isn't ALL the ram on the RPI DMA coherent?  It's just a matter of getting the mappings correct, right?

**Comment by ghost** [comment]
I mean, couldn't the memory be allocated with kmalloc or friends, then mapped into DMA space with dma_map_single/dma_map_sg/friends when it's actually sent to the USB controller?

**Comment by clivem** [comment]
I'll leave the technical discussion to the experts..... 
I've now migrated 18 instances of various Pi hardware to 4.8.y HEAD with MULTIPLATFORM config and "coherent_pool=1M". I'll keep an eye on the logs.......

**Comment by P33M** [comment]
Coherent mappings require no explicit cache flushing between hardware changing values in RAM and the CPU reading them and vice versa. Bits are set in the page entries that effectively bypass cache on these pages.

dwc_otg is suboptimal in that regard - an unaligned memcpy from coherent memory is going to be a bit slow, but the majority of transfers that fall into the "unaligned" bucket are of the order of 16 bytes. 

dwc2 upstream allocates an aligned buffer and overwrites urb->transfer_buffer to allow the usb-related DMA map-unmap functions to work, but frees/allocates on submission and retire of each URB with unaligned buffers. I don't know which would be slower on a Pi 1 - the coherent unaligned memcpy or the allocate/deallocate for each URB.

**Comment by popcornmix** [comment]
@clivem can you identify the USB device that results in the dwc_otg allocations?
Either unplug one USB device at a time and see if number of `caller=__DWC_DMA_ALLOC_ATOMIC` lines reduces, or add a `WARN` to the current logging to get a backtrace. Interested to know if you have a single device that results in 4 x 64K allocations, or you have a number of devices with one or two each.

**Comment by clivem** [comment]
@popcornmix Only USB device plugged into that board (which generated the earlier log I posted) is an Edimax wifi dongle using the mt7601u kernel module.

```
Feb 11 16:28:02 MamboLSPlus2B1 kernel: usb 1-1.5: new high-speed USB device number 4 using dwc_otg
Feb 11 16:28:02 MamboLSPlus2B1 kernel: usb 1-1.5: New USB device found, idVendor=7392, idProduct=7710
Feb 11 16:28:02 MamboLSPlus2B1 kernel: usb 1-1.5: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Feb 11 16:28:09 MamboLSPlus2B1 kernel: usb 1-1.5: reset high-speed USB device number 4 using dwc_otg
Feb 11 16:28:09 MamboLSPlus2B1 kernel: mt7601u 1-1.5:1.0: ASIC revision: 76010001 MAC revision: 76010500
Feb 11 16:28:09 MamboLSPlus2B1 kernel: mt7601u 1-1.5:1.0: Firmware Version: 0.1.00 Build: 7640 Build time: 201302052146____
Feb 11 16:28:09 MamboLSPlus2B1 kernel: mt7601u 1-1.5:1.0: Warning: unsupported EEPROM version 0d
Feb 11 16:28:09 MamboLSPlus2B1 kernel: mt7601u 1-1.5:1.0: EEPROM ver:0d fae:00
Feb 11 16:28:09 MamboLSPlus2B1 kernel: mt7601u 1-1.5:1.0: EEPROM country region 01 (channels 1-13)
Feb 11 16:28:10 MamboLSPlus2B1 kernel: usbcore: registered new interface driver mt7601u
```

**Comment by popcornmix** [comment]
Okay, I've bumped default coherent size to 1M and added a WARN_ON to failure path when allocating a coherent buffer from dwc_otg, so we might catch users with multiple USB devices that use unaligned transfers.

**Comment by clivem** [comment]
@popcornmix Cool! Any chance of rebase 4.8.4 -> 4.8.5 before I run another local build. ;)

**Comment by popcornmix** [comment]
Rebased. 4.8.6 now.

**Comment by clivem** [comment]
@popcornmix Ta.

**Pull request merged**